src/rand.c

```diff
@@ -23,12 +23,20 @@
 #include <string.h> /* memeset */
 #include <time.h>   /* clock, time */
 
+#if !defined(DONT_USE_GETRANDOM) && defined(__linux__) && \
```

Comment:
I don't have strong objections against your approach. Just pay attention that getrandom() can block according to the manual.

Maybe we can improve DRBG implementation by using SHA512. So systems other than linux will get faster.

Reply:
This file contains `xmpp_alloc()` call. Try to build with --disable-getrandom to see the errors.

Reply:
> Just pay attention that getrandom() can block according to the manual.

True: "If the urandom source has not yet been initialized", i.e. very early in the kernel start phase, do you think this can be a real problem? I'm pretty sure you're having other parts also failing that early in the boot phase ...

> This file contains xmpp_alloc() call. Try to build with --disable-getrandom to see the errors.

Yeah, realized already but patch was mixed with the change to introduce `strophe_alloc`

Reply:
I tried SHA512 instead of SHA1 and it works much faster for big chunks since it generates 64 bytes at once instead of 20. But it slightly worse for 2-16 blocks where single SHA1 iteration is faster than SHA512 iteration. Most of the time, generation spends in `sha512_compress()`.

This can be improved further, there are different variant of the NIST DRBG, also there is fast BLAKE3 hash function.

Reply:
I played a bit around with it too, but as the results were similar to the results with SHA1 and nothing ground-breaking, I didn't spend more time on it.

Those are really interesting problems, but this is a XMPP library and no crypto library ... we should focus on that :)

Also the really hard part isn't about making the PRNG fast, but about making it secure. I.e. before spending time into using other/faster algorithms, we should improve on the seeding!

TBH I think the best would be to use the internal PRNG only as last resort! I know, it's less fun, but still :)

1. OS specific way (`getrandom`, `arc4random`)
2. when TLS is enabled: use the RNG from the TLS stack, as it is usually secure, fast and easy to use.
3. internal PRNG either with software hash or when TLS is enabled maybe also think about using the hash API's provided by the TLS stack (they're most likely way faster than our software hash as they use an optimized version to the platform we're running on)

IMO we should stop where we're now and if someone comes and says "but but but I need this and that" re-visit the problem.

Reply:
Agree, you can mark this as resolved.